diagnose_cluster.py

```diff
@@ -0,0 +1,119 @@
+parser.add_argument('--overwrite', required=False, action='store_true', help="Delete dest directory before adding new files")
+parser.add_argument('--no-diagnose', required=False, action='store_true', help="Do not run gcloud dataproc clusters diagnose")
+parser.add_argument('--compress', '-z', required=False, action='store_true', help="GZIP all files")
+parser.add_argument('--workers', required=False, nargs='*', help="Specific workers to get log files from")
+parser.add_argument('--take', required=False, type=int, default=None, help="Only download logs from the first N workers")
+
+
+args = parser.parse_args()
+
+is_local = not args.dest.startswith("gs://")
+
+if args.overwrite:
+    if is_local:
+        call('rm -r {dir}'.format(dir=args.dest), shell=True)
+    else:
+        call('gsutil -m rm -r {dir}'.format(dir=args.dest), shell=True)
+    
+
+master_dest = args.dest.rstrip('/') + "/master/"
+worker_dest = args.dest.rstrip('/') + "/workers/"
+
+if is_local:
+    call('mkdir -p {dir}'.format(dir=master_dest), shell=True)
+    call('mkdir -p {dir}'.format(dir=worker_dest), shell=True)
+
+desc = json.loads(Popen('gcloud dataproc clusters describe {name} --format json'.format(name=args.name),
+                        shell=True,
+                        stdout=PIPE,
+                        stderr=PIPE).communicate()[0].strip())
+
+config = desc['config']
+
+master = config['masterConfig']['instanceNames'][0]
+workers = config['workerConfig']['instanceNames']
+zone = re.search('zones\/(?P<zone>\S+)$', config['gceClusterConfig']['zoneUri']).group('zone')
+
+if args.workers:
+    invalid_workers = set(args.workers).difference(set(workers))
+    assert len(invalid_workers) == 0, "Non-existent workers specified: " + ", ".join(invalid_workers)
+    workers = args.workers
+
+if args.take:
+    assert args.take > 0 and args.take <= len(workers), "Number of workers to take must be in the range of [0, nWorkers]. Found " + args.take + "."
+    workers = workers[:args.take]
+
+    
+def gcloud_ssh(remote, command):
+    return 'gcloud compute ssh {remote} --zone {zone} --command "{command}"'.format(remote=remote, zone=zone, command=command)
+
+
+def gcloud_copy_files(remote, src, dest):
+    return 'gcloud compute copy-files {remote}:{src} {dest} --zone {zone}'.format(remote=remote, src=src, dest=dest, zone=zone)
+
+
+def gsutil_cp(src, dest):
+    return 'gsutil -m cp -r {src} {dest}'.format(src=src, dest=dest)
+
+
+def copy_files_tmp(remote, files, dest, tmp):
+    init_cmd = ['mkdir -p {tmp}; rm -r {tmp}/*'.format(tmp=tmp)]
+    
+    copy_tmp_cmds = ['sudo cp -r {file} {tmp}'.format(file=file, tmp=tmp) for file in files]
+    copy_tmp_cmds.append('sudo chmod -R 777 {tmp}'.format(tmp=tmp))
+
+    if args.compress:
+        copy_tmp_cmds.append('sudo find ' + tmp + ' -type f ! -name \'*.gz\' -exec gzip "{}" \;')
+    
+    call(gcloud_ssh(remote, '; '.join(init_cmd + copy_tmp_cmds)), shell=True)
```

Comment:
Really all of the `shell=True`, even if it doesn't seem like you could possibly do something wrong. Better safe